Set cxx standard, reconfigure root

SimonRit opened the pull request:
I have been struggling to compile the develop branch of gate due to issues with cxx standard of root. I have now implemented a more meaningful message but that enforces the same cxx standard or Geant4, Gate and root. If you like it, it would be great to check that it works before merging, in particular the torch part. Thanks!

Commit:
Document CMAKE_POLICY changes for readability

Commit:
Use Geant4 cxx standard (currently >=17)

Commit:
Update ROOT management in cmake files

thisroot.sh now sets CMAKE_PREFIX_PATH and the latest FindROOT.cmake can
be used. This allows proper definition of root targets.

Commit:
Setting cxx standard for Torch is useless since Geant uses at least 17

tbaudier commented:
The modifications seems good for me.
I just want to adapt the Github actions workflow:
https://github.com/OpenGATE/Gate/blob/develop/.github/workflows/main.yml#L80-L166

If I understand well:
 - root need c++17 flag
 - Geant4 v11 does not need c++17 flag anymore
 - Gate does not need c++17 flag anymore
 - libtorch need at least c++14
 - RTK need c++17 flag ?

I will try to push the modification on you branch to see the behavior
Thank you

SimonRit commented:
> root need c++17 flag

My suggestion is to force the user to use the same as Geant4 (so c++17, c++20 or c++23, see below)


> Geant4 v11 does not need c++17 flag anymore

Geant4 only allows 17, 20 and 23 (if available):
https://gitlab.cern.ch/geant4/geant4/-/blob/master/cmake/Modules/G4BuildSettings.cmake#L164-L168

> Gate does not need c++17 flag anymore

Idem as root, I set it equal to Geant4's. 

> libtorch need at least c++14

I don't know if one can use different C++ versions....

> RTK need c++17 flag ?

It compiles with c++14 but I haven't tested it yet...

tbaudier committed:
Adapt c++ flags in github actions

Now Gate is compiled with the same flag than Geant and root.
So I removed the c++17 flag in Geant because it's the default behavior and in Gate because Gate copy the flag from Geant
I do not remove the flag from RTK and root compilation because there is no other way to know the needed c++ flag
I change the cache name to be sure to recompile root and geant4
